### PR TITLE
gin-config: init

### PR DESCRIPTION
**Pull request by jethrokuan**
###### Motivation for this change
Add new python package `gin-config`

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/python-modules/gin-config/default.nix
Comment (Ma27) [inline pkgs/development/python-modules/gin-config/default.nix]: I'd add an optional parameter `withTensorflow ? false` here as it seems to be optional according to `setup.py`: https://github.com/google/gin-config/blob/master/setup.py#L51

This might look like this then:

``` nix
{
  # ...
  propagatedBuildInputs = [ six enum34 ] ++ lib.optional withTensorflow tensorflow;
}
```

From my understanding this project doesn't need to be used for ML projects and in such a case I doubt that you want to have `tensorflow` in your build closure :)

**Review by Ma27 — CHANGES_REQUESTED** [review]
Except my comment regarding `tensorflow` I'm :+1:

File: pkgs/development/python-modules/gin-config/default.nix
Comment (Ma27) [inline pkgs/development/python-modules/gin-config/default.nix]: this should be `++`.
`lib.optional` does basically the following: `cond: a: if cond then [ a ] else []`.

**Comment by Ma27** [comment]
have you tested this on Python3?
I get the following failure when building from your branch:

```
/nix/store/9pifap75pkb9ifmdqh0irn8wgc3q38s6-stdenv-linux/setup: line 1298:  1786 Illegal instruction     /nix/store/bwfygfcdvis9wd1c1v51xwnwhw1hx0a0-python3-3.6.6/bin/python3.6m nix_run_setup test
```

Let's see what borg says about this:

@GrahamcOfBorg build pythonPackages.gin-config python3Packages.gin-config

**Comment by Ma27** [comment]
ahh right... should've pushed the fix at first :D

@GrahamcOfBorg build pythonPackages.gin-config python3Packages.gin-config

File: pkgs/development/python-modules/gin-config/default.nix
Comment (jethrokuan) [reply pkgs/development/python-modules/gin-config/default.nix]: oops thanks! amateur mistake :P

**Comment by jethrokuan** [comment]
Works for me.

```
hash-lda2vec on  master [⇡!?] 
➜ python                                                                                                 1s 289ms → 2 
Python 3.6.6 (default, Jun 27 2018, 05:47:41) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import gin
>>> 
```

**Comment by Ma27** [comment]
yeah, that's what I suspected, unfortunately. I obvserved such issues some times in recent history, but I'm not sure yet how to proceed here.

Let's see if I have sufficient time to debug this tomorrow, if I forget about this, feel free to ping me :)

**Comment by jethrokuan** [comment]
fwiw i built with this command `nix-build -I nixpkgs=/home/jethro/nix-config/nixpkgs/ -A python36Packages.gin-config`, let me know if I should be running something else

File: pkgs/development/python-modules/gin-config/default.nix
Comment (FRidh) [inline pkgs/development/python-modules/gin-config/default.nix]: This is pointless. If one (does not) want Tensorflow, they will need to override this expression. In that case they might as well use `overridePythonAttrs`. Tensorflow is an optional dependency, so if someone wants to use this package in conjunction with tensorflow, they can just add both to the environment.

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/gin-config/default.nix
Comment (FRidh) [reply pkgs/development/python-modules/gin-config/default.nix]: I wrote a separate comment already, but no, there should not be an option for it.

File: pkgs/development/python-modules/gin-config/default.nix
Comment (Ma27) [reply pkgs/development/python-modules/gin-config/default.nix]: you're right, sorry, missed that during my review :/

Comment (jethrokuan) [reply pkgs/development/python-modules/gin-config/default.nix]: I made changes to the derivation based on what I _think_ you mean, hope it's correct!

**Comment by Ma27** [comment]
@jethrokuan what @FRidh wanted to say is that `tensorflow` isn't needed as propagated build input. It's optional (as I mentioned), if you need it, you can add this to your python environment as well.

Sorry, I didn't think of this, that was clearly my mistake!
Regarding the Python3 issue, I'll have a deeper look into this tomorrow ok? :)

**Comment by Ma27** [comment]
So, I suspect that tensorflow seems to have caused the issue locally. Removing it from the list of check inputs starts at least the test suite.

However, the PyPI source archive doesn't ship the tests, so now we can do two things:

* Use `fetchFromGitHub` instead and get the tests running
* Simply add `doCheck = false;` (several python modules do this)

Except for that issue, the PR is looking good now and should be ready to merge :)

**Comment by FRidh** [comment]
When enabling tests, I suggest disabling tests that require tensorflow because it's a very heavy dependency.

**Comment by jethrokuan** [comment]
> However, the PyPI source archive doesn't ship the tests.

You're right, totally didn't catch this. I'll add `doCheck = false;`, and make a note about it.

**Comment by jethrokuan** [comment]
I couldn't build it locally with `propogatedBuildInputs`, it would attempt to install `six` and fail.

```
building '/nix/store/hd5zykp5kcn33qy8m6bc2s33cg0xp873-python3.6-gin-config-0.1.1.2.drv'...
unpacking sources
unpacking source archive /nix/store/y4d9jkfa89dd0b2z2bkkdfwlclb8l127-gin-config-0.1.1.2.tar.gz
source root is gin-config-0.1.1.2
setting SOURCE_DATE_EPOCH to timestamp 1531854765 of file gin-config-0.1.1.2/setup.cfg
patching sources
configuring
building
['six >= 1.10.0']
running bdist_wheel
running build
running build_py
creating build
creating build/lib
creating build/lib/gin
copying gin/utils.py -> build/lib/gin
copying gin/__init__.py -> build/lib/gin
copying gin/selector_map.py -> build/lib/gin
copying gin/config.py -> build/lib/gin
copying gin/config_parser.py -> build/lib/gin
creating build/lib/gin/tf
copying gin/tf/external_configurables.py -> build/lib/gin/tf
copying gin/tf/utils.py -> build/lib/gin/tf
copying gin/tf/__init__.py -> build/lib/gin/tf
running egg_info
writing gin_config.egg-info/PKG-INFO
writing dependency_links to gin_config.egg-info/dependency_links.txt
writing requirements to gin_config.egg-info/requires.txt
writing top-level names to gin_config.egg-info/top_level.txt
reading manifest file 'gin_config.egg-info/SOURCES.txt'
writing manifest file 'gin_config.egg-info/SOURCES.txt'
installing to build/bdist.linux-x86_64/wheel
running install
running install_lib
creating build/bdist.linux-x86_64
creating build/bdist.linux-x86_64/wheel
creating build/bdist.linux-x86_64/wheel/gin
copying build/lib/gin/utils.py -> build/bdist.linux-x86_64/wheel/gin
creating build/bdist.linux-x86_64/wheel/gin/tf
copying build/lib/gin/tf/external_configurables.py -> build/bdist.linux-x86_64/wheel/gin/tf
copying build/lib/gin/tf/utils.py -> build/bdist.linux-x86_64/wheel/gin/tf
copying build/lib/gin/tf/__init__.py -> build/bdist.linux-x86_64/wheel/gin/tf
copying build/lib/gin/__init__.py -> build/bdist.linux-x86_64/wheel/gin
copying build/lib/gin/selector_map.py -> build/bdist.linux-x86_64/wheel/gin
copying build/lib/gin/config.py -> build/bdist.linux-x86_64/wheel/gin
copying build/lib/gin/config_parser.py -> build/bdist.linux-x86_64/wheel/gin
running install_egg_info
Copying gin_config.egg-info to build/bdist.linux-x86_64/wheel/gin_config-0.1.1.2-py3.6.egg-info
running install_scripts
creating build/bdist.linux-x86_64/wheel/gin_config-0.1.1.2.dist-info/WHEEL
creating 'dist/gin_config-0.1.1.2-py3-none-any.whl' and adding 'build/bdist.linux-x86_64/wheel' to it
adding 'gin/__init__.py'
adding 'gin/config.py'
adding 'gin/config_parser.py'
adding 'gin/selector_map.py'
adding 'gin/utils.py'
adding 'gin/tf/__init__.py'
adding 'gin/tf/external_configurables.py'
adding 'gin/tf/utils.py'
adding 'gin_config-0.1.1.2.dist-info/METADATA'
adding 'gin_config-0.1.1.2.dist-info/WHEEL'
adding 'gin_config-0.1.1.2.dist-info/top_level.txt'
adding 'gin_config-0.1.1.2.dist-info/RECORD'
removing build/bdist.linux-x86_64/wheel
installing
/build/gin-config-0.1.1.2/dist /build/gin-config-0.1.1.2
Processing ./gin_config-0.1.1.2-py3-none-any.whl
Collecting six>=1.10.0 (from gin-config==0.1.1.2)
  Could not find a version that satisfies the requirement six>=1.10.0 (from gin-config==0.1.1.2) (from versions: )
No matching distribution found for six>=1.10.0 (from gin-config==0.1.1.2)
builder for '/nix/store/hd5zykp5kcn33qy8m6bc2s33cg0xp873-python3.6-gin-config-0.1.1.2.drv' failed with exit code 1
cannot build derivation '/nix/store/bx3jacj0jyzkg8dlsyh6sh4rsyxan985-python3-3.6.6-env.drv': 1 dependencies couldn't be built
error: build of '/nix/store/bx3jacj0jyzkg8dlsyh6sh4rsyxan985-python3-3.6.6-env.drv' failed
```

**Comment by Ma27** [comment]
may I ask from which state you built this? I just checked out your branch, built `gin-config` and I don't have `/nix/store/hd5zykp5kcn33qy8m6bc2s33cg0xp873-python3.6-gin-config-0.1.1.2.drv` in my nix store.

Furthermore, please change `buildInputs` back to `propagatedBuildInputs`, then it works again on runtime :)

**Comment by Ma27** [comment]
@GrahamcOfBorg build pythonPackages.gin-config python3Packages.gin-config

**Review by Ma27 — APPROVED** [review]
Checked the basic functionality locally. As most of our Python packages have zero tests (due to fetching the PyPI tarball) this should be fine for now :+1:

**Comment by Ma27** [comment]
@FRidh as you requested changes here as well, I'd wait for your approval ok? :)

File: pkgs/development/python-modules/gin-config/default.nix
Comment (jethrokuan) [inline pkgs/development/python-modules/gin-config/default.nix]: xD oops

**Comment by FRidh** [comment]
Ah, the commits were not squashed together :(

**Pull request merged**
